packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+    } else {
+        Vec::new()
+    };
+
+    let mut children2 = component
+        .get_rendered_children()
+        .iter()
+        .filter_map(|child| match child {
+            &ComponentChild::Component(comp_idx) => Some(comp_idx),
+            _ => None,
+        })
+        .collect();
+
+    children.append(&mut children2);
+
+    children
+}
+
+/// If the state variable specified by original_component_state is stale or unresolved,
+/// then freshen the variable, resolving its dependencies if necessary.
+///
+/// If the state variable was not fresh, then recurse to its dependencies to freshen them.
+pub fn freshen_state_var(
+    original_component_state: ComponentStateDescription,
+    components: &Vec<Rc<RefCell<ComponentEnum>>>,
+    dependencies: &mut Vec<Vec<Vec<Vec<Dependency>>>>,
+    dependent_on_state_var: &mut Vec<Vec<Vec<ComponentStateDescription>>>,
+    dependent_on_essential: &mut Vec<HashMap<EssentialDataOrigin, Vec<ComponentStateDescription>>>,
+    essential_data: &mut Vec<HashMap<EssentialDataOrigin, EssentialStateVar>>,
+    freshen_stack: &mut Vec<StateVarCalculationState>,
+    should_initialize_essential_data: bool,
+) -> () {
+    // This function currently implements recursion through an iterative method,
+    // using a stack on the heap.
+    // This approach was chosen because the function recursion implementation would overflow
+    // the small WASM stack once it recursed a few thousands of times.
+    // An alternative approach would be to increase the size of the WASM stack.
+    // This approach was chosen because it is relatively easy to produce documents
+    // with thousands of levels in the dependency graph, and it wasn't clear what
+    // size WASM stack would be appropriate.
+
+    let current_freshness = components[original_component_state.component_idx]
+        .borrow()
+        .get_state_variables()[original_component_state.state_var_idx]
+        .get_freshness();
+
+    let initial_calculation_state = match current_freshness {
+        // No need to do anything if the state var is already fresh
+        Freshness::Fresh => return (),
+        Freshness::Unresolved => StateVarCalculationState::Unresolved(UnresolvedCalculationState {
+            component_state: original_component_state,
+            instruction_idx: 0,
+            val_idx: 0,
+            dependency_instructions: None,
+            instruct_dependencies: None,
+            dependencies_for_state_var: None,
+        }),
+        Freshness::Stale => StateVarCalculationState::Stale(StaleCalculationState {
+            component_state: original_component_state.clone(),
+            instruction_idx: 0,
+            val_idx: 0,
+        }),
+    };
+
+    freshen_stack.push(initial_calculation_state);
+
+    'stack_loop: while let Some(calculation_state) = freshen_stack.pop() {
+        match calculation_state {
+            StateVarCalculationState::Unresolved(unresolved_state) => {
```

Comment:
I've been trying to read this branch, but it confuses me...Could you instead just create the dep, set it as stale, and repush it to the stack and let the other branch handle it?

Reply:
Yes, I theoretically could put it back on the stack as stale and let the other branch handle it. But that would only save me the last two lines (given that the dependencies are fresh at this point):

```
state_var.calculate_state_var_from_dependencies();
state_var.record_all_dependencies_viewed();
```

Reply:
Perhaps some of this could be refactored into another function, then? The code is quite hard to follow...

Reply:
refactored code, so resolving this conversation.